### PR TITLE
fix(compile): improve handling of Windows junctions and symbolic links

### LINKED ISSUE #26316: `Deno compile` errors when ancestor directory is windows junction with target on separate drive
Version:  2.0.0-rc.9

_**Note this happens when I have a legacy program with a package.json** Removing package.json is a simpler fix but needs a better diagnostic to tell me that.

I have an app in y:\x\bin, but that's really a link to C:\Users\Bob\OneDrive\xfer\bin. This confuses the compiler. It should be smart about handling junctions in Windows.

```
Y:\x\Bin\xform>deno compile --allow-read --allow-write --output xform.exe xform.ts
Check file:///Y:/x/Bin/xform/xform.ts
Compile file:///Y:/x/Bin/xform/xform.ts to xform.exe
error: Writing temporary file 'xform.exe.tmp-7f879bfc5a53ae10'

Caused by:
    Failed to strip prefix 'C:\Users\Bob\OneDrive\xfer\bin' from 'Y:\x\Bin\xform\package.json'
```

the work-around is a .cmd file which moves to the real directory and does the compiler there.

### PR DESCRIPTION
**Pull request by yazan-abdalrahman**
fix #26316

### CODE REVIEW

**Comment by yazan-abdalrahman** [comment]
@dsherret Please can review this change.

File: cli/standalone/virtual_fs.rs [hunk 1]
```diff
@@ -184,11 +183,61 @@ impl VfsBuilder {
...
   pub fn add_file_at_path(&mut self, path: &Path) -> Result<(), AnyError> {
-    let target_path = canonicalize_path(path)?;
-    if target_path != path {
-      self.add_symlink(path, &target_path)?;
+    let canonicalized_target_path = canonicalize_path(path)?;
+    self.resolve_symlinks_in_path(path, &canonicalized_target_path)?;
+    self.add_file_at_path_not_symlink(&canonicalized_target_path)
+  }
+
+  fn resolve_symlinks_in_path(
```
Comment (dsherret) [inline cli/standalone/virtual_fs.rs:191]: Can you explain what this function does? I'm having trouble following it.

File: cli/standalone/virtual_fs.rs [hunk 2]
```diff
@@ -184,11 +183,61 @@ impl VfsBuilder {
...
+    &mut self,
+    path: &Path,
+    canonicalized_target_path: &PathBuf,
+  ) -> Result<(), AnyError> {
+    #[cfg(windows)]
```
Comment (dsherret) [inline cli/standalone/virtual_fs.rs:196]: Why should we only do this for Windows?

**Review by dsherret — CHANGES_REQUESTED** [review]
I've been trying to reproduce this and I'm not sure this is the issue. Have you been able to create a reproduction? Is there a test you know of that fails on main?

**Comment by dsherret** [comment]
I investigate this one and it's already fixed on main via the changes in (https://github.com/denoland/deno/pull/26494 which surfaced this bug more readily).

That said, I don't believe this is the right fix for this issue and I don't understand the code in the PR (I'm not sure why it's so complex). Instead of doing this, we can instead ignore adding a symlink/junction when the newpath is outside the root directory (which is what #26494 did). Maybe in the future we could have some way to have a root that is above multiple windows drives, but for the time being having the restriction that all the files must live under a link to a different drive is not too bad.

**Comment by yazan-abdalrahman** [comment]
> I investigate this one and it's already fixed on main via the changes in (#26494 which surfaced this bug more readily).
> 
> That said, I don't believe this is the right fix for this issue and I don't understand the code in the PR (I'm not sure why it's so complex). Instead of doing this, we can instead ignore adding a symlink/junction when the newpath is outside the root directory (which is what #26494 did). Maybe in the future we could have some way to have a root that is above multiple windows drives, but for the time being having the restriction that all the files must live under a link to a different drive is not too bad.

THX,
Sure, ignoring symlink is easier.


In this scenario, my method attempts to canonicalize the path until it matches the canonicalized root. However, if there are multiple symlinks, the root can be canonicalized until the last symlink, which can cause issues. Therefore, this dynamic method resolves each symlink one by one until the root and path match.

**Pull request closed without merging**